File: python/pbase/typefun/cloud.py
import numpy as np
from typing import Tuple, List, Optional

from .. import mathc
from .. import types

from .. import lib

# // Prints the whole cloud data to stdout
# void p_cloud_print(pCloud self);
lib.p_cloud_print.argtypes = [types.pCloud]


def cloud_print(self: np.ndarray):
    '''
    Prints the whole cloud data to stdout
    '''
    lib.p_cloud_print(types.cast_np_pCloud(self))


# // Concatenates two point clouds together
# pCloud p_cloud_concatenate(pCloud a, pCloud b);
lib.p_cloud_concatenate.argtypes = [types.pCloud,
                                    types.pCloud]
lib.p_cloud_concatenate.restype = types.pCloud


def cloud_concatenate(a: np.ndarray,
                      b: np.ndarray) \
        -> np.ndarray:
    '''
    Concatenates two point clouds together
    
    :return: pCloud
    '''
    res = lib.p_cloud_concatenate(types.cast_np_pCloud(a),
                                  types.cast_np_pCloud(b))
    return types.cast_pCloud_np(res)


# // Concatenates a list/vector of point clouds together
# pCloud p_cloud_concatenate_v(const pCloud *cloud_list, int n);
lib.p_cloud_concatenate_v.argtypes = [types.pCloud_p,
                                      types.c_int]
lib.p_cloud_concatenate_v.restype = types.pCloud


def cloud_concatenate_v(cloud_list: List[np.ndarray]) \
        -> np.ndarray:
    '''
    Concatenates a list/vector of point clouds together
    
    :return: pCloud
    '''
    n = len(cloud_list)
    LIST = types.pCloud * n
    list = LIST()
    for i in range(n):
        list[i] = types.cast_np_pCloud(cloud_list[i])

    res = lib.p_cloud_concatenate_v(list,
                                    n)
    return types.cast_pCloud_np(res)


# // Applies indices on cloud, so that all not used points are removed
# pCloud p_cloud_apply_indices(pCloud cloud, pIndices indices);
lib.p_cloud_apply_indices.argtypes = [types.pCloud,
                                      types.pIndices]
lib.p_cloud_apply_indices.restype = types.pCloud


def cloud_apply_indices(self: np.ndarray,
                        indices: np.ndarray) \
        -> np.ndarray:
    '''
    Applies indices on cloud, so that all not used points are removed

    :return: pCloud
    '''
    res = lib.p_cloud_apply_indices(types.cast_np_pCloud(self),
                                    types.cast_np_pIndices(indices))
    return types.cast_pCloud_np(res)

File: src/typefun/cloud.c
#include <stdio.h>
#include <string.h>

#include "pbase/mathc/sca/int.h"
#include "pbase/typefun/cloud.h"


void p_cloud_print(pCloud self) {
    if(!p_cloud_valid(self)) {
        puts("p_cloud_print: invalid");
        return;
    }
    puts("p_cloud_print:");
    for (int i = 0; i < self.size; i++) {
        printf("%.2f %.2f %.2f %.2f\n",
               self.data[i].x,
               self.data[i].y,
               self.data[i].z,
               self.data[i].w
        );
    }
}

pCloud p_cloud_concatenate(pCloud a, pCloud b) {
    return p_cloud_concatenate_v((pCloud[]) {a, b}, 2);
}

pCloud p_cloud_concatenate_v(const pCloud *cloud_list, int n) {
    size_t size = 0;
    for(int i=0; i<n; i++)
        size += cloud_list[i].size;
    pCloud self = p_cloud_new_empty(size);
    vec4 *data = self.data;
    for(int i=0; i<n; i++) {
        size = cloud_list[i].size;
        memcpy(data, cloud_list[i].data, size * sizeof *data);
        data += size;
    }
    return self;
}

pCloud p_cloud_apply_indices(pCloud self, pIndices indices) {
    pCloud ret = p_cloud_new_empty(indices.size);
    for(int i=0; i<indices.size; i++) {
        int index = isca_mod_positive(indices.data[i], self.size);
        ret.data[i] = self.data[index];
    }
    return ret;
}
